File: modules/ark-speech/android/src/main/java/expo/modules/arkspeech/ArkSpeechModule.kt
package expo.modules.arkspeech

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.speech.RecognitionListener
import android.speech.RecognizerIntent
import android.speech.SpeechRecognizer
import androidx.core.content.ContextCompat
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.Locale

class ArkSpeechModule : Module() {
  private val mainHandler = Handler(Looper.getMainLooper())
  private var recognizer: SpeechRecognizer? = null
  private var activePromise: Promise? = null
  private var timeoutRunnable: Runnable? = null

  override fun definition() = ModuleDefinition {
    Name("ArkSpeech")

    AsyncFunction("isAvailable") { promise: Promise ->
      val context = appContext.reactContext
      promise.resolve(context != null && SpeechRecognizer.isRecognitionAvailable(context))
    }

    AsyncFunction("recognizeOnce") { options: Map<String, Any?>?, promise: Promise ->
      val context = appContext.reactContext
        ?: return@AsyncFunction promise.reject(
          "ERR_CONTEXT_UNAVAILABLE",
          "Android context is not available for speech recognition.",
          null
        )

      if (ContextCompat.checkSelfPermission(context, Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
        promise.reject("ERR_MIC_PERMISSION", "Microphone permission is required for voice input.", null)
        return@AsyncFunction
      }

      if (!SpeechRecognizer.isRecognitionAvailable(context)) {
        promise.reject("ERR_SPEECH_UNAVAILABLE", "Android speech recognition is not available on this device.", null)
        return@AsyncFunction
      }

      if (activePromise != null) {
        promise.reject("ERR_SPEECH_BUSY", "Speech recognition is already active.", null)
        return@AsyncFunction
      }

      activePromise = promise
      val preferOffline = options?.get("preferOffline") as? Boolean ?: true
      val locale = options?.get("locale") as? String ?: Locale.getDefault().toLanguageTag()
      val timeoutMs = ((options?.get("timeoutMs") as? Number)?.toLong() ?: 45000L).coerceIn(5000L, 120000L)

      mainHandler.post {
        val nextRecognizer = SpeechRecognizer.createSpeechRecognizer(context)
        recognizer = nextRecognizer
        nextRecognizer.setRecognitionListener(object : RecognitionListener {
          override fun onReadyForSpeech(params: Bundle?) = Unit
          override fun onBeginningOfSpeech() = Unit
          override fun onRmsChanged(rmsdB: Float) = Unit
          override fun onBufferReceived(buffer: ByteArray?) = Unit
          override fun onEndOfSpeech() = Unit
          override fun onPartialResults(partialResults: Bundle?) = Unit
          override fun onEvent(eventType: Int, params: Bundle?) = Unit

          override fun onError(error: Int) {
            rejectActive("ERR_SPEECH_FAILED", speechErrorMessage(error), null)
          }

          override fun onResults(results: Bundle?) {
            val matches = results
              ?.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION)
              ?.map { it.trim() }
              ?.filter { it.isNotEmpty() }
              ?: emptyList()
            val text = matches.firstOrNull().orEmpty()
            resolveActive(
              mapOf(
                "text" to text,
                "alternatives" to matches,
                "offlinePreferred" to preferOffline
              )
            )
          }
        })

        val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).apply {
          putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM)
          putExtra(RecognizerIntent.EXTRA_PARTIAL_RESULTS, false)
          putExtra(RecognizerIntent.EXTRA_PREFER_OFFLINE, preferOffline)
          putExtra(RecognizerIntent.EXTRA_LANGUAGE, locale)
          putExtra(RecognizerIntent.EXTRA_MAX_RESULTS, 3)
        }

        timeoutRunnable = Runnable {
          recognizer?.cancel()
          rejectActive("ERR_SPEECH_TIMEOUT", "No speech was recognized before the timeout.", null)
        }
        mainHandler.postDelayed(timeoutRunnable!!, timeoutMs)
        nextRecognizer.startListening(intent)
      }
    }

    Function("stop") {
      mainHandler.post {
        recognizer?.stopListening()
      }
    }

    Function("cancel") {
      mainHandler.post {
        recognizer?.cancel()
        rejectActive("ERR_SPEECH_CANCELLED", "Speech recognition was cancelled.", null)
      }
    }
  }

  private fun resolveActive(value: Any) {
    val promise = activePromise ?: return
    cleanup()
    promise.resolve(value)
  }

  private fun rejectActive(code: String, message: String, cause: Throwable?) {
    val promise = activePromise ?: return
    cleanup()
    promise.reject(code, message, cause)
  }

  private fun cleanup() {
    timeoutRunnable?.let { mainHandler.removeCallbacks(it) }
    timeoutRunnable = null
    recognizer?.destroy()
    recognizer = null
    activePromise = null
  }

  private fun speechErrorMessage(error: Int): String {
    return when (error) {
      SpeechRecognizer.ERROR_AUDIO -> "Android could not capture microphone audio."
      SpeechRecognizer.ERROR_CLIENT -> "Android speech recognition was interrupted."
      SpeechRecognizer.ERROR_INSUFFICIENT_PERMISSIONS -> "Microphone permission is required for voice input."
      SpeechRecognizer.ERROR_NETWORK -> "Offline speech recognition is unavailable for this language on this device."
      SpeechRecognizer.ERROR_NETWORK_TIMEOUT -> "Speech recognition timed out while waiting for the recognizer."
      SpeechRecognizer.ERROR_NO_MATCH -> "No speech was recognized."
      SpeechRecognizer.ERROR_RECOGNIZER_BUSY -> "Android speech recognition is already busy."
      SpeechRecognizer.ERROR_SERVER -> "Android speech recognition service failed."
      SpeechRecognizer.ERROR_SPEECH_TIMEOUT -> "No speech was detected."
      else -> "Android speech recognition failed."
    }
  }
}

File: modules/ark-speech/ios/ArkSpeechModule.swift
import AVFoundation
import ExpoModulesCore
import Speech

public class ArkSpeechModule: Module {
  private let audioEngine = AVAudioEngine()
  private var recognitionRequest: SFSpeechAudioBufferRecognitionRequest?
  private var recognitionTask: SFSpeechRecognitionTask?
  private var activePromise: Promise?
  private var timeoutWorkItem: DispatchWorkItem?

  public func definition() -> ModuleDefinition {
    Name("ArkSpeech")

    AsyncFunction("isAvailable") { () -> Bool in
      return SFSpeechRecognizer()?.isAvailable ?? false
    }

    AsyncFunction("recognizeOnce") { (options: [String: Any]?, promise: Promise) in
      self.recognizeOnce(options: options, promise: promise)
    }.runOnQueue(.main)

    Function("stop") {
      DispatchQueue.main.async {
        self.audioEngine.stop()
        self.recognitionRequest?.endAudio()
      }
    }

    Function("cancel") {
      DispatchQueue.main.async {
        self.rejectActive(Exception(name: "ERR_SPEECH_CANCELLED", description: "Speech recognition was cancelled."))
      }
    }
  }

  private func recognizeOnce(options: [String: Any]?, promise: Promise) {
    guard activePromise == nil else {
      promise.reject(Exception(name: "ERR_SPEECH_BUSY", description: "Speech recognition is already active."))
      return
    }

    activePromise = promise
    requestSpeechAuthorization { [weak self] authorized in
      guard let self else {
        return
      }
      guard authorized else {
        self.rejectActive(Exception(name: "ERR_SPEECH_PERMISSION", description: "Speech recognition permission is required for voice input."))
        return
      }
      AVAudioSession.sharedInstance().requestRecordPermission { granted in
        DispatchQueue.main.async {
          guard granted else {
            self.rejectActive(Exception(name: "ERR_MIC_PERMISSION", description: "Microphone permission is required for voice input."))
            return
          }
          self.startRecognition(options: options)
        }
      }
    }
  }

  private func startRecognition(options: [String: Any]?) {
    let localeIdentifier = options?["locale"] as? String ?? Locale.current.identifier
    let preferOffline = options?["preferOffline"] as? Bool ?? true
    let timeoutMs = min(max((options?["timeoutMs"] as? NSNumber)?.doubleValue ?? 45_000, 5_000), 120_000)
    let recognizer = SFSpeechRecognizer(locale: Locale(identifier: localeIdentifier)) ?? SFSpeechRecognizer()

    guard let recognizer, recognizer.isAvailable else {
      rejectActive(Exception(name: "ERR_SPEECH_UNAVAILABLE", description: "iOS speech recognition is not available on this device."))
      return
    }

    if preferOffline, !recognizer.supportsOnDeviceRecognition {
      rejectActive(Exception(name: "ERR_SPEECH_OFFLINE_UNAVAILABLE", description: "Offline iOS speech recognition is unavailable for this language on this device."))
      return
    }

    let request = SFSpeechAudioBufferRecognitionRequest()
    request.shouldReportPartialResults = false
    if #available(iOS 13.0, *) {
      request.requiresOnDeviceRecognition = preferOffline
    }

    do {
      let audioSession = AVAudioSession.sharedInstance()
      try audioSession.setCategory(.record, mode: .measurement, options: .duckOthers)
      try audioSession.setActive(true, options: .notifyOthersOnDeactivation)
    } catch {
      rejectActive(Exception(name: "ERR_AUDIO_SESSION", description: error.localizedDescription))
      return
    }

    recognitionRequest = request
    let inputNode = audioEngine.inputNode
    let format = inputNode.outputFormat(forBus: 0)
    inputNode.removeTap(onBus: 0)
    inputNode.installTap(onBus: 0, bufferSize: 1024, format: format) { buffer, _ in
      request.append(buffer)
    }

    audioEngine.prepare()
    do {
      try audioEngine.start()
    } catch {
      rejectActive(Exception(name: "ERR_AUDIO_START", description: error.localizedDescription))
      return
    }

    recognitionTask = recognizer.recognitionTask(with: request) { [weak self] result, error in
      guard let self else {
        return
      }
      if let result, result.isFinal {
        let text = result.bestTranscription.formattedString.trimmingCharacters(in: .whitespacesAndNewlines)
        self.resolveActive([
          "text": text,
          "alternatives": text.isEmpty ? [] : [text],
          "offlinePreferred": preferOffline
        ])
        return
      }
      if let error {
        self.rejectActive(Exception(name: "ERR_SPEECH_FAILED", description: error.localizedDescription))
      }
    }

    let timeout = DispatchWorkItem { [weak self] in
      self?.rejectActive(Exception(name: "ERR_SPEECH_TIMEOUT", description: "No speech was recognized before the timeout."))
    }
    timeoutWorkItem = timeout
    DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(Int(timeoutMs)), execute: timeout)
  }

  private func requestSpeechAuthorization(_ completion: @escaping (Bool) -> Void) {
    SFSpeechRecognizer.requestAuthorization { status in
      DispatchQueue.main.async {
        completion(status == .authorized)
      }
    }
  }

  private func resolveActive(_ value: Any) {
    guard let promise = activePromise else {
      return
    }
    cleanup()
    promise.resolve(value)
  }

  private func rejectActive(_ error: Exception) {
    guard let promise = activePromise else {
      return
    }
    cleanup()
    promise.reject(error)
  }

  private func cleanup() {
    timeoutWorkItem?.cancel()
    timeoutWorkItem = nil
    audioEngine.inputNode.removeTap(onBus: 0)
    if audioEngine.isRunning {
      audioEngine.stop()
    }
    recognitionRequest?.endAudio()
    recognitionTask?.cancel()
    recognitionTask = nil
    recognitionRequest = nil
    activePromise = nil
    try? AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
  }
}
